[게시판 회원 기능] 제이 미션 제출합니다

JunHoPark93 opened the pull request:
항상 감사드립니다!

MrKwon committed:
docs: Readme 업데이트

MrKwon committed:
feat: 게시글 생성, 목록 조회 기능 구현

MrKwon committed:
feat: 게시글 조회 구현

MrKwon committed:
feat: 공통 Head 분리

JunHoPark93 committed:
feat: html 중복제거

JunHoPark93 committed:
feat: 게시글 수정

JunHoPark93 committed:
feat: 게시글 삭제 구현

JunHoPark93 committed:
feat: devtools 적용

JunHoPark93 committed:
refactor: Js 분리

JunHoPark93 committed:
refactor: 컨트롤러 메서드 이름 수정, autowired 삭제

JunHoPark93 committed:
refactor: 불필요한 메서드 제거

JunHoPark93 committed:
feat: Article 제목 검증 추가

JunHoPark93 committed:
feat: error 페이지 추가

JunHoPark93 committed:
refactor: 피드백 반영

리다이렉트, 이름 잘못 수정된것 원복, header location 검증 추가

JunHoPark93 committed:
feat: Jpa repository 적용

JunHoPark93 committed:
Merge branch 'temp' into JunHoPark93

# Conflicts:
#	README.md
#	build.gradle
#	src/main/java/techcourse/myblog/domain/Article.java
#	src/main/java/techcourse/myblog/web/ArticleController.java
#	src/main/resources/application.properties
#	src/main/resources/templates/article.html
#	src/test/java/techcourse/myblog/web/ArticleControllerTests.java

JunHoPark93 committed:
docs: README 업데이트

JunHoPark93 committed:
feat: 회원 등록 기능

JunHoPark93 committed:
feat: 유저 조회 기능 구현

Squashed commit of the following:

commit 3026e5c7eb726b65b992756d9ec17a73bec140d4
Author: JunHoPark <pjh374968@gmail.com>
Date:   Tue Jul 16 17:54:01 2019 +0900

    feat: 유저 조회 기능 구현

commit c82f723cc643fd9aed7686d0231ed18cd35924b7
Author: JunHoPark <pjh374968@gmail.com>
Date:   Tue Jul 16 16:53:26 2019 +0900

    fix: Valid 애노테이션 추가

    Validation 에러 수정, @Valid가 없으면 검증이 발동되지 않는다

JunHoPark93 committed:
feat: 로그인 및 로그아웃 기능 구현

Squashed commit of the following:

commit 3af160545dc0fcdce403342652fd1c9497e60982
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:57:01 2019 +0900

    refactor: html 패키지화 원복

commit abaa172ae7507edcc0a8a21852291f5b216d967a
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:53:36 2019 +0900

    feat: 로그아웃 기능 구현

commit a523c371383247a05158711cdc7a1d1c6cd1c2ff
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:26:37 2019 +0900

    feat: 로그인 기능 구현

commit 010778b87f6995e0d38a29d15b3041529cb43b20
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 15:25:59 2019 +0900

    chore: JPA 실행 쿼리 보기 설정

commit fcf46a795cfdb3dcdd14a683d93dd1c1ea0817b7
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 14:48:54 2019 +0900

    feat: 세션 검증 Interceptor 구현

commit 261149e005613cc1fe6ab4ead0485842ff80780c
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 13:12:06 2019 +0900

    refactor: html 패키지 구조 변경

JunHoPark93 committed:
feat: 회원 수정 및 탈퇴기능 구현

Squashed commit of the following:

commit d36b93b9f56d5a631168ea59425740e3852b4c23
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 17:29:53 2019 +0900

    feat: 회원 탈퇴 기능 구현

commit 0a0adfd710c88ca96f77b0a14b71812556c0643d
Author: JunHoPark <pjh374968@gmail.com>
Date:   Wed Jul 17 17:07:59 2019 +0900

    feat: 회원 수정 기능 구현

JunHoPark93 committed:
refactor: html 중복제거 및 세션적용

JunHoPark93 committed:
fix: Interceptor에서 /users post요청은 통과

JunHoPark93 committed:
feat: Error 객체의 정의

Login 관련 에러 처리 객체

JunHoPark93 committed:
test: 세션관련 인수테스트 작성

JunHoPark93 committed:
test: 회원관련 인수테스트

JunHoPark93 committed:
refactor: 들여쓰기 및 사용하지 않는 import 제거

JunHoPark93 committed:
feat: 회원가입시 이메일 중복 오류 처리

JunHoPark93 committed:
fix: 회원 가입시 email 중복 테스트 수정

JunHoPark93 committed:
refactor: UserService 분리

JunHoPark93 committed:
feat: 회원 수정시 이름 형식 오류 처리

JunHoPark93 committed:
feat: 회원 가입시 유효하지 않은 입력 처리

JunHoPark93 committed:
refactor: Article 서비스 패키지 분리

JunHoPark93 committed:
fix: 유저 리스트에서 다른 유저 정보에 접근하는 오류 처리

JunHoPark93 committed:
feat: 데이터베이스 구축

* 프로덕션 환경은 mysql, 테스트 환경은 h2를 사용하였다.

JunHoPark93 committed:
refactor: 삼항 연산자로 변경

* login, logout을 구분하는 if-else문을 삼항 연산자로 변경

* article의 기본 그림을 구분하는 if-else문을 삼항 연산자로 변경

JunHoPark93 committed:
refactor: 삼항 연산자로 변경

JunHoPark93 committed:
fix: 게시글 id 접근하는 로직 수정

JunHoPark93 committed:
fix: 절대 경로로 변경

JunHoPark93 committed:
refactor: 삼항 연산자로 변경, url 변경

JunHoPark93 committed:
refactor: fragments 패키지로 분리

woowahan-pjs reviewed (changes requested):
오프라인 코드 리뷰로 대신하였습니다. 👍

JunHoPark93 committed:
refactor: 피드백 반영

메서드 이름 수정
패키지 구조 전체적으로 변경 (root 기준으로 좀 더 깔끔해 지게 변경, 레이어드 아키텍쳐를 고려하여 depth 설정)
JPA 실행 쿼리 보여주기
JPA repository의 사용
Domain을 생성하는 곳 통일 (dto에서 domain 생성 vs 서비스에서 domain 생성 -> 서비스로 통일)

JunHoPark93 commented:
# 피드백 반영 내용 

코치님 말씀들을 반영해본 내용들입니다.

## 패키지 구조 변경

* 패키지 구조를 root에서 봤을 때 조금 더 보기 쉽게 변경. 예를들어 인터셉터나 validation 같은 경우는 web 패키지 밑에 들어간다. 에러 관련 클래스들은 컨트롤러에서 쓰지만 서비스 패키지 밑에 두어도 상관없다. 레이어드 아키텍쳐관점에서 상위 것들은 밑에 것들을 알 수 있기 때문에
* 편의성을 위해서 CrudRepositoy 보다는 JpaRepository를 사용해보기.
* 설정파일이나 핸들러 클래스들을 support 패키지로 분리.
* 도메인 생성 부분은 취향차이 : dto에서 도메인 생성 vs 서비스에서 도메인 생성 -> 저는 서비스에서 도메인생성하는 것이 나아보여서 서비스에서 도메인생성으로 통일하였습니다.
* JPA 로깅 추가
* 기타 내용들 반영

JunHoPark93 committed:
feat: 비밀번호 암호화 구현

BCrypt 라이브러리를 이용한다. 내 프로젝트에 맞게 wrapping 작업을 한다. 그 이유는 크게 두가지이다.
1. BCrypt 클래스의 이름과 내부 메서드 철자의 어려움
2. 프로젝트의 특성에 맞게 새로 클래스를 구현할 수 있음 (인터페이스로 분리하여 다른 암호화를 쉽게 Bean을 통해 주입받아 변경 가능)

마지막으로 비밀번호 검증은 Web Config 와 연관있다고 생각하여 interceptor를 구현한 클래스를 수정하여 Web 설정과 관련된 @Bean들의 집합 클래스로 변경

JunHoPark93 committed:
refactor: implements 삭제

JunHoPark93 commented:
## 추가구현 내용

### 비밀번호 암호화
그냥 암호에 plain text를 넣는게 좋지않아서 암호화를 적용시켜보았습니다.
말씀하신것처럼 시큐리티를 쓰는게 지금 구현해볼 단계는 아니라고 생각해서 직접 암호화 라이브러리를 이용하여 wrapping을 시켜 저희 프로젝트에 적용시켜보았습니다.

피드백 이후 변경 내용입니다.

jbcrypt라는 아주 기능이 간단한 라이브러리를 추가하였고 클래스를 하나 더 만들어서 그것을 감쌌습니다.
추상화 인터페이스 - 나중에 암호화 방식이 바꾸더라도 구현체만 갈아끼워주면 되기 때문에 추상화를 해보았습니다.
```java
public interface EncryptHelper {
    String encrypt(String password);

    boolean isMatch(String password, String hashed);
}
```

인터페이스를 이용한 구현체를 만들었습니다. 내부적으로는 BCrypt 클래스를 이용하고있습니다.

```java
import org.mindrot.jbcrypt.BCrypt;

public class BCryptImpl implements EncryptHelper {
    @Override
    public String encrypt(String password) {
        return BCrypt.hashpw(password, BCrypt.gensalt());
    }

    @Override
    public boolean isMatch(String password, String hashed) {
        return BCrypt.checkpw(password, hashed);
    }
}

```
제가 그렇게 한 이유는,
1. BCrypt 클래스의 이름과 내부 메서드 철자의 어려움
2. 프로젝트의 특성에 맞게 새로 클래스를 구현할 수 있음 (인터페이스로 분리하여 다른 암호화를 쉽게 Bean을 통해 주입받아 변경 가능 → 유연한 구조를 만들어볼 수 있을 것 같았음)

혹시 틀렸다면 지적부탁드립니다!

그리고 인터셉터를 활용한 WebMvcConfigurer를 구현한 클래스가 있었는데 비밀번호 암호화도 Web에 관련된 것이라 생각해서 구조를 바꾸었습니다.

기존구조
```java
@Configuration
public class WebConfig implements WebMvcConfigurer {
    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(new AuthInterceptor())
                .addPathPatterns("/users")
                .addPathPatterns("/mypage-edit")
                .addPathPatterns("/mypage")
                .addPathPatterns("/mypage/*");
    }
}
```

바꾼 구조
```java
@Configuration
public class WebConfig {
    @Bean
    public WebMvcConfigurer interceptorConfigure() {
        return new WebMvcConfigurer() {
            @Override
            public void addInterceptors(InterceptorRegistry registry) {
                registry.addInterceptor(new AuthInterceptor())
                        .addPathPatterns("/users")
                        .addPathPatterns("/mypage-edit")
                        .addPathPatterns("/mypage")
                        .addPathPatterns("/mypage/*");
            }
        };
    }

    @Bean
    public EncryptHelper encryptConfigure() {
        return new BCryptImpl();
    }
}
```
정의한 구현체를 빈으로 등록하여 전체적인 웹 설정 클래스로 바꿔보았습니다. WebMVCConfigurer 설정을 커스텀하여 사용하는 부분들이 많아진다면 다시 분리하는 것도 나쁘지 않을것 같습니다. 지금은 설정을 쓰는게 많이 없어서 한 군데 에 몰아넣어보았습니다.

그리고 서비스에서 주입을 받아서 사용했습니다.
```java
@Service
public class UserService {
    private UserRepository userRepository;
    private EncryptHelper encryptHelper;

    public UserService(UserRepository userRepository, EncryptHelper encryptHelper) {
        this.userRepository = userRepository;
        this.encryptHelper = encryptHelper;
    }
```

고민한 내용이 암호화관련된 클래스를 도메인 (User)에서 가지고 있는것은 아닌것 같아서 서비스에서 사용을 하는데 해시된 비밀번호를 검증하는 과정에서 도메인의 password를 가져오는 부분이 있습니다. 

```java
public User findUserByEmail(UserLoginRequest userLoginRequest) {
        User user = userRepository.findUserByEmail(userLoginRequest.getEmail())
                .orElseThrow(() -> new LoginException("email 없음"));

        checkPassword(userLoginRequest, user);
        return user;
    }

    private void checkPassword(UserLoginRequest userLoginRequest, User user) {
        if (!encryptHelper.isMatch(userLoginRequest.getPassword(), user.getPassword())) {
            throw new LoginException("비밀번호 틀림");
        }
    }
```
도메인의 isPasswordMatch 함수를 쓰는 것보다(그냥 단순 해시값 비교) encrypter의 라이브러리를 사용하기 위해서 get을 했는데 (그리고 encrypter에 비교하는 처리를 넘겨줌) 이 부분이 혹시 이상한지 궁금합니다. 도메인에서도 해시된 비밀번호를 가지고 있어서 get을 해도 별 문제가 없을 것 같기도해서요.

JunHoPark93 committed:
feat: 비밀번호, 이메일 검증 로직 개선

JunHoPark93 committed:
feat: 검증 로직 추가

비밀번호, 비밀번호 확인 필드 검증 기능
회원 수정 dto 추가

JunHoPark93 committed:
fix: 없는 게시물 접근 처리

JunHoPark93 committed:
test: 테스트 파라미터 추가

JunHoPark93 committed:
fix: email validation 추가

JunHoPark93 committed:
test: cookie를 이용한 테스트

woowahan-pjs reviewed (changes requested):
오프라인 코드 리뷰로 대신하였습니다. 👍

JunHoPark93 committed:
refactor: 피드백 반영

생성자 private, location 검증 수정

JunHoPark93 committed:
refactor: localport 제거

JunHoPark93 committed:
refactor: 패키지 이름 변경

JunHoPark93 committed:
refactor: 클래스 이름 변경

Impl 네이밍은 불필요하다. 조금 더 암호화 구체 클래스의 기능을 나타내는 이름으로 변경하였다.

JunHoPark93 commented:
피드백 감사합니다! 알려주신 링크 잘 읽어보겠습니다!

woowahan-pjs reviewed (approved):
스프링의 세계에 온 것을 환영합니다. 👍 
이만 Merge 하겠습니다.
